### PR TITLE
Bump plutus dep to 1.16

### PR DESCRIPTION
**Pull request by michaelpj**
Routine bump, lets us get rid of some instances, and supports newer `nothunks`.

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -18,9 +18,9 @@ source-repository-package
 
 index-state:
   -- Bump this if you need newer packages from Hackage
-  , hackage.haskell.org 2023-11-09T23:50:15Z
+  , hackage.haskell.org 2023-11-15T00:00:00Z
```
Comment (jasagredo) [inline cabal.project:21]: I think the ledger team prefers to set this to an existing index state, to avoid getting the cabal warning.

Comment (michaelpj) [reply cabal.project:21]: It's a time in the past, isn't that sufficient?

Comment (jasagredo) [reply cabal.project:21]: Nope
```
Warning: Requested index-state 2023-11-15T00:00:00Z is newer than
'hackage.haskell.org'! Falling back to older state (2023-11-14T11:26:13Z).
Warning: Requested index-state 2023-11-15T00:00:00Z is newer than
'cardano-haskell-packages'! Falling back to older state
(2023-11-14T12:07:29Z).
```

Comment (jasagredo) [reply cabal.project:21]: Note this warning will be downgraded when this PR gets into the released Cabal https://github.com/haskell/cabal/pull/8944 but for now a warning is emitted.

Comment (lehins) [reply cabal.project:21]: If I understand correctly this whole nix/cabal index thing `2023-11-15T00:00:00Z` is the correct time:
https://github.com/input-output-hk/hackage.nix/blob/f7760ad0bc4014d7fb0786326d14842e9ebd0641/index-state-hashes.nix#L1691

Comment (lehins) [reply cabal.project:21]: I guess I was wrong. Then I definitely do not understand it :see_no_evil:

**Review by lehins — APPROVED** [review]
Looks good to me. I am not gonna worry about the cabal warning.

File: cabal.project (hunk 1 above)
Comment (jasagredo) [reply cabal.project:21]: https://github.com/input-output-hk/cardano-ledger/pull/3871#pullrequestreview-1731920543

Let's resolve this then

**Comment by lehins** [comment]
@michaelpj We prefer rebase instead of commits that merge master into ledger

This results in a nice linear history

I've fixed this in this PR by merging this PR with a squashing, but just letting you know for the future.

**Comment by michaelpj** [comment]
Ah right, I just hit the "update branch" button because it was there!

**Pull request merged**